libnmstate/nm/context.py

```diff
@@ -62,12 +100,98 @@ def _del_client(self):
                 timeout_source = GLib.timeout_source_new(50)
                 try:
                     timeout_source.set_callback(lambda x: is_timeout.append(1))
-                    timeout_source.attach(self._context)
+                    timeout_source.attach(self.context)
                     while not is_done and not is_timeout:
-                        self._context.iteration(True)
+                        self.context.iteration(True)
                 finally:
                     timeout_source.destroy()
-
             if not is_done:
                 logging.error("BUG: NM.Client is not cleaned")
             self._context = None
+
+    def _del_timeout(self):
+        if self._timeout_source:
+            self._timeout_source.destroy()
+            self._timeout_source = None
+
+    def register_async(self, action, fast=False):
+        """
+        Register action(string) to wait list.
+        """
+        queue = self._fast_queue if fast else self._slow_queue
+        max_queue = FAST_ASYNC_QUEUE_SIZE if fast else SLOW_ASYNC_QUEUE_SIZE
+        if len(queue) >= max_queue:
+            logging.debug(
+                f"Async queue({max_queue}) full, waiting all existing actions "
+                "to be finished before registering more async action"
+            )
+            # TODO: No need to wait all finish, a single finished async is
+            #       sufficient.
+            self.wait_all_finish()
+
+        logging.debug(f"Async action: {action} started")
+        queue.add(action)
+
+    def finish_async(self, action, suppress_log=False):
+        """
+        Mark action(string) as finished.
+        """
+        self._last_async_finish_time = datetime.datetime.now()
+        if not suppress_log:
+            logging.debug(f"Async action: {action} finished")
+        self._fast_queue.discard(action)
+        self._slow_queue.discard(action)
+
+    def _action_all_finished(self):
```

Comment:
This may represent that all actions finished or that no action started.
How about `number_of_registered_actions`? Or `registered_actions`?

Reply:
If there is no action register then wait all finish should return immediately. I don't see a big changes made by this renaming.

Reply:
If you do not differentiate between the two scenarios, it is less important.
IMO it is misleading, but wont insist on this either.